crates/turbopack-ecmascript/src/references/dynamic_expression.rs

```diff
@@ -0,0 +1,71 @@
+use anyhow::Result;
+use serde::{Deserialize, Serialize};
+use swc_core::quote;
+use turbo_tasks::{debug::ValueDebugFormat, trace::TraceRawVcs, Vc};
+
+use super::AstPath;
+use crate::{
+    chunk::EcmascriptChunkingContext,
+    code_gen::{CodeGenerateable, CodeGeneration},
+    create_visitor,
+};
+
+#[derive(PartialEq, Eq, TraceRawVcs, Serialize, Deserialize, ValueDebugFormat)]
+enum DynamicExpressionType {
+    Promise,
+    Normal,
+}
+
+#[turbo_tasks::value]
+pub struct DynamicExpression {
+    path: Vc<AstPath>,
+    ty: DynamicExpressionType,
+}
+
+#[turbo_tasks::value_impl]
+impl DynamicExpression {
+    #[turbo_tasks::function]
+    pub fn new(path: Vc<AstPath>) -> Vc<Self> {
+        Self::cell(DynamicExpression {
+            path,
+            ty: DynamicExpressionType::Normal,
+        })
+    }
+
+    #[turbo_tasks::function]
+    pub fn new_promise(path: Vc<AstPath>) -> Vc<Self> {
+        Self::cell(DynamicExpression {
+            path,
+            ty: DynamicExpressionType::Promise,
+        })
+    }
+}
+
+#[turbo_tasks::value_impl]
+impl CodeGenerateable for DynamicExpression {
+    #[turbo_tasks::function]
+    async fn code_generation(
+        &self,
+        _context: Vc<Box<dyn EcmascriptChunkingContext>>,
+    ) -> Result<Vc<CodeGeneration>> {
+        let path = &self.path.await?;
+
+        let visitor = match self.ty {
+            DynamicExpressionType::Normal => {
+                create_visitor!(path, visit_mut_expr(expr: &mut Expr) {
+                    *expr = quote!("(() => { const e = new Error(\"Cannot find module as expression is too dynamic\"); e.code = 'MODULE_NOT_FOUND'; throw e; })()" as Expr);
```

Comment:
this will completely throw away information about what would have been called

can we add something back so it shows up in the stack trace?

Reply:
Giving the inline function a name would probably work

Reply:
Due to source mapping the error will point to the actual user code.

Reply:
But the stack trace still won't contain a `require`/`import`